Wallbox Integration

bullitt186 opened the pull request:
As discussed in https://github.com/torbennehmer/hacs-e3dc/issues/47, i implemented the Wallbox features of python-e3dc 0.9.2.

I could test the features on my S10X compact with a MultiConnect II:
- Sensor readings work and are plausible
- Sun Charging can be set
- Phases can be toggled
- Max charging current can be set via service

I wasn't able to validate
- Schuko (as the Multiconnect II does not have one)
- Toggle Charging (as my car was fully charged today)

As of now, only 1 Wallbox is supported.
As more than one Wallbox requires some entity name juggeling, i skipped this for now.
I had to add the button entity to this integration for the toggles.

I am happy for testers, suggestions and improvements.

bullitt186 commented:
Hi Torben, Thanks for the positive feedback 👍
I am curious to see the changes you did but find only your comment here but no other commits or change request to the pull request. Am i missing something, looking in the wrong places or did you integrate it locally on your end and will push later?

Please let me know if you would like to see any other changes from my end before integrating the PR.

by the way in the meantime i was able to verify that the toggle for "Wallbox Charging" works as expected.

Cheer,
Bastian

torbennehmer commented:
Ah Bastian, I'm very sorry, that got lost in translation.
I did not change anything, I added comments to the PR by using a PR review, you should see them in the changes view of the PR.
Cheers,
Torben

bullitt186 commented:
Hmmm, That's odd. But maybe it's also a layer 8 problem on my end.
I cannot find your remarks anywhere.
I would have expected to see such a "changes requested" mark as shown in the screenshot or as you said in the changes view. In my last PR it also was like that.
https://docs.github.com/assets/cb-157392/mw-1440/images/help/pull_requests/merge_box/pr-reviews-in-merge-box.webp

Could you please doublechecked whether you submitted your PR review?

ps - while searching for your comments, i realized my local PyLinter changed more formatting as I intended and this must have missed my attention. sorry. will fix this if neccesary.

torbennehmer commented:
> Hmmm, That's odd. But maybe it's also a layer 8 problem on my end. I cannot find your remarks anywhere. [...]
> Could you please doublechecked whether you submitted your PR review?

Nah, it was my Layer 8 :-) Klicking on "Submit review" helped, I guess. Sorry for that.

> ps - while searching for your comments, i realized my local PyLinter changed more formatting as I intended and this must have missed my attention. sorry. will fix this if neccesary.

Yeah, thought that. If you have an idea how to configure the VSCode Workspace to have this consistently across all dev systems, I'd be happy to take an independant ticket and PR for this. I don't stick to a specific style of formatting, it just should be consistent, so that we don't reformat back and forth between us :) This probably has to include line lenghts, linter and formatter types etc., which would have to be set in the workspace config, not in the user config. I'm no expert for the options we've for vscode Python development, to be honest, so I'd be happy for input on this end.

Cheers, Torben

bullitt186 commented:
Ah yes, now i have your feedback to work with :-) Thank you.
Let's see whether i find the time tonight to get it done.

Regarding linting: I used my lunch break to explore this a little bit. I put my thoughts/results in this issue: https://github.com/torbennehmer/hacs-e3dc/issues/154
Let's continue the discussion on linting there.

bullitt186 commented on .vscode/settings.json:
```diff
@@ -3,5 +3,8 @@
-    "python.formatting.provider": "black"
+    "python.formatting.provider": "black",
+    "files.associations": {
+        "*.yaml": "home-assistant"
```
Hmm... Tp be honest i don't know how it came here...

bullitt186 commented on custom_components/e3dc_rscp/strings.json:
```diff
@@ -189,10 +207,40 @@
       },
       "wallbox-total": {
-        "name": "Wallbox - total"
+        "name": "Wallbox Wallbox - total"
```
Some copy/paste glitch --> reverted

bullitt186 commented on custom_components/e3dc_rscp/strings.json:
```diff
@@ -189,10 +207,40 @@
       },
       "farm-additional-total": {
-        "name": "Farm additional powermeter - total"
+        "name": "Wallbox Farm additional powermeter - total"
```
Some copy/paste glicht

bullitt186 commented on .gitignore:
done

bullitt186 commented on .devcontainer.json:
```diff
@@ -1,6 +1,6 @@
 {
     "name": "torbennehmer/hacs-e3dc",
-    "image": "mcr.microsoft.com/vscode/devcontainers/python:0-3.11",
+    "image": "mcr.microsoft.com/devcontainers/python:3.12",
```
i got an error from a HA dependency (cannot recall which one) that Python 3.12 is needed.
But yeah, let's not mix up 2 different things

bullitt186 commented on custom_components/e3dc_rscp/e3dc_proxy.py:
```diff
@@ -226,6 +237,96 @@ def start_manual_charge(self, charge_amount_wh: int) -> None:
+            raise HomeAssistantError("Failed to toggle wallbox charging")
+
+    @e3dc_call
+    def toggle_wallbox_phases(self, wbIndex: int = 0):
```
To my knowledge ony 1+3 phases are supported, thus a simple toggle is okay to switch between two states.
(That's what's being provided by the python-e3dc lib)
I am not a big fan of toggles as they are stateless but there's a delay between toggeling and reading back the phases via "wallbox_phases", so changing to a switch is no real option.

bullitt186 commented on custom_components/e3dc_rscp/e3dc_proxy.py:
```diff
@@ -226,6 +237,96 @@ def start_manual_charge(self, charge_amount_wh: int) -> None:
+
+        """
+
+        MAX_CHARGE_CURRENT = 32  # Maximum allowed current in Amperes
```
done

bullitt186 commented on custom_components/e3dc_rscp/services.py:
```diff
@@ -114,6 +133,22 @@ def _resolve_device_id(hass: HomeAssistant, devid: str) -> E3DCCoordinator:
+    hass: HomeAssistant, call: ServiceCall
+) -> None:
+    """Extract service information and relay to coordinator."""
+    # TODO: Addoption to select Wallbox
```
Add option was meant.
This will be relevant as soon as more than one wallbox is supported

neuromancer3142 commented:
Hi guys,

Any way I can support here wrt. end user testing? I do not see the new E3DC version in HACS for testing yet. Please let me know when I can test from my end. I do have an S10E Pro with an E3DC Wallbox Easy Connect.

Cheers.

bullitt186 commented on custom_components/e3dc_rscp/button.py:
done

bullitt186 replied:
done

bullitt186 commented on custom_components/e3dc_rscp/coordinator.py:
done

bullitt186 commented on custom_components/e3dc_rscp/translations/en.json:
done

bullitt186 commented on custom_components/e3dc_rscp/coordinator.py:
done

bullitt186 commented on custom_components/e3dc_rscp/binary_sensor.py:
done

bullitt186 commented on custom_components/e3dc_rscp/binary_sensor.py:
done

bullitt186 commented on custom_components/e3dc_rscp/sensor.py:
done

bullitt186 commented on custom_components/e3dc_rscp/sensor.py:
```diff
@@ -384,6 +469,7 @@ async def async_setup_entry(
     entities: list[E3DCSensor] = [
         E3DCSensor(coordinator, description, entry.unique_id)
         for description in SENSOR_DESCRIPTIONS
+        if coordinator._wallbox_installed or not description.key.startswith("wallbox-")
```
done; introduced proper getter/setter

bullitt186 commented on custom_components/e3dc_rscp/strings.json:
done

bullitt186 commented on custom_components/e3dc_rscp/switch.py:
done

bullitt186 commented on custom_components/e3dc_rscp/switch.py:
done

bullitt186 commented on custom_components/e3dc_rscp/e3dc_proxy.py:
done

bullitt186 commented on .DS_Store:
done, removed file from repo

neuromancer3142 commented:
> @neuromancer3142 - in order to find these changes in HACS, they need to be fully integrated + released first. if you want to test them in it's current state already, you'd have to clone my fork (https://github.com/bullitt186/hacs-e3dc/tree/wallbox) or download it's [zip](https://github.com/bullitt186/hacs-e3dc/archive/refs/heads/wallbox.zip) and place the content of custom_components in your config folder of HA. (or open the devcontainer in vscode and then run /scripts/develop which spins up a local instance of Home Assistant)
> 
> I'd be curious to get some feedback.
> 

Happy to provide feedback. Easiest way for me would be to have it in one of the next releases and then give it a go. If that is not happening anytime soon, I will look into the download part. I fear of breaking my HA instance and not being able to fix it ;-).

Cheers.

torbennehmer commented:
> Happy to provide feedback. Easiest way for me would be to have it in one of the next releases and then give it a go. If that is not happening anytime soon, I will look into the download part. I fear of breaking my HA instance and not being able to fix it ;-).

We can push out a Beta release, I'll look into the PR today and check that.

neuromancer3142 commented:
Thanks a lot Torben ...

torbennehmer commented on custom_components/e3dc_rscp/sensor.py:
```diff
@@ -372,6 +372,92 @@
         device_class=SensorDeviceClass.POWER_FACTOR,
         state_class=SensorStateClass.MEASUREMENT,
     ),
+    SensorEntityDescription(
```
Could you set this as 
`entity_category=EntityCategory.DIAGNOSTIC,`
it'll appear along the other diagnostics information at the bottom of the device page. See at the top of the file for more examples. Sorry, missed that in the review.

bullitt186 replied:
Good point, yes will do.
(in fact it does only report "0" - i found in a discussions somewhere ( think at python-e3dc) that this seems to be normal)

torbennehmer commented on custom_components/e3dc_rscp/sensor.py:
```diff
@@ -372,6 +372,92 @@
+        device_class=SensorDeviceClass.ENERGY,
+        state_class=SensorStateClass.TOTAL_INCREASING,
+    ),
+    SensorEntityDescription(
```
Would this be also a candidate for a diagnostics entity?

bullitt186 replied:
Yes, will add this.

bullitt186 replied:
yes, changed that

bullitt186 commented:
I think i found the cause for the odd switch behaviour, am currently working on it. so you may want to hold back your efforts for the beta release for a little bit.

torbennehmer commented:
Hi @bullitt186,

> [...] There's one thing i don't get... In the clip below, you see the switches fall back to "unknown" but i can't figure out why. The switches appear identical to the others of the extension (e.g. config switch weather regulation) which acts differently. The switch is not updated by me somewhere else. Just to be clear: the state of the wallbox (e.g. sun mode) is changed correctly and does not fall back, but HA "forgets" the switch state. @torbennehmer - do you have any ideas what may cause this behaviour?

I have seen this behavior before, yes. What I saw with weather regulated charging was about this:

1. User clicks the switch, it flips
2. Code behind this pings E3DC to update the setting
3. Meanwhile, E3DC updates the entity based on the coordinator's state, which still has the old(!) value of the setting. This is because you're racing against the next E3DC poll.
4. At the next poll (10s later max.) the integration polls E3DC again and then updates to the new setting.

I had to work around this in the coordinator to actually update the state immediately when I'm sending service commands to E3DC. Basically, you have to optimistically update the integrations internal state, hoping that the next poll will actually match this. Look at the following line for an example.

https://github.com/torbennehmer/hacs-e3dc/blob/31b13491399217c11f72dbdaed95c3b24ffb1017/custom_components/e3dc_rscp/coordinator.py#L326

Another example:

https://github.com/torbennehmer/hacs-e3dc/blob/31b13491399217c11f72dbdaed95c3b24ffb1017/custom_components/e3dc_rscp/coordinator.py#L340

You'd still have a last edge case in case a poll is already going on while you're temporarily patching the coordinator state. I haven't seen that so far, but in theory that's possible as we're running multi-threaded.

That's why I'm using a guard variable to protect against this. I'm not sure if this is a true Python synchronisation style, but it did work so far. Any insights from a Python expert in multi-threading would be appreciated here :)

https://github.com/torbennehmer/hacs-e3dc/blob/31b13491399217c11f72dbdaed95c3b24ffb1017/custom_components/e3dc_rscp/coordinator.py#L338

Cheers, Torben

bullitt186 commented:
> That's why I'm using a guard variable to protect against this. I'm not sure if this is a true Python synchronisation style, but it did work so far. Any insights from a Python expert in multi-threading would be appreciated here :)

I guess as well there's a more native way to handle mutex but for now I copied your style.

With the latest commits i did some more refinement, in summary:

- added mutex behaviour
- made switch and state to be in sync (with the limitations stated by @torbennehmer above)
- changed the binay_sensors to native HA classes, this made it necessary to rename the entity names
- realized i had some strings/translations at the wrong entity type and corrected this.
-  set the entities which are specify to WB type multi connect I (schuko) and easy connect (plug locked state) to disabled by default in order not to confuse users. I haven't found a way to identify the WB type, ideally the user could select their WB type during set up of the integration
- more testing next to the car to verify the states / behaviour

In accordance to the quote "don't let perfect be in the way of better" i think this is ready to release to early adopters.

torbennehmer commented:
cool thing, I'll go over it right away.

Thomansky commented:
I have the problem that only the settings for a wallbox are set for me and the battery score does not work for me from the car either, it is permanently displayed at 100%.

bullitt186 commented:
Hi @Thomansky 
Thank you very much for your initial feedback.
In fact, as of now all functions only work with the primary wall box or to be more precise for the Wallbox with the lowest ID. 

I would expect that it is consistent throughout all sensors and switches that only one of the wall boxes information are shown or settings can be set. There should not be a mix between different wallboxes.

Supporting multiple wall boxes is on the to do list but probably won't make it into the initial release.


Regarding state of charge: I guess this depends on the communication between the specific vehicle and the wall box.
In my case I have a Mercedes EQB and a multi connect 2 wall box and state of charge is always zero.

I will investigate this a little bit.

Thomansky commented:
Okay, I've just connected my car. Battery level 96% in the Home Assistant is currently displayed at 3% but every 30 seconds it increases by 1%, I'll wait until it reaches 96% maybe it's just an update problem.

Thomansky commented:
Okay, I've just connected my car. Battery level 96% in the Home Assistant is currently displayed at 3% but every 30 seconds it increases by 1%, I'll wait until it reaches 96% maybe it's just an update problem.

Thomansky commented:
The car's charge level actually seems to be updated by 1% every 30 seconds until the value in Home Assistant is at the car's current value

Thomansky commented:
Is it not possible to choose which wallbox is prioritised?

bullitt186 commented:
@Thomansky Within the HA integration, i just pass the values received by the Hauskraftwerk via a python library (https://github.com/fsantini/python-e3dc)  through and am not altering any values. Maybe this is the system behaviour of the hauskraftwerk learning the SoC from the vehicle?
Did you plug in the car just before checking the values?

Wallbox prioritization is  not implemented in this library and i haven't seen this feature in E3DCs RSCP spec, so I'm afraid this is not available. However maybe you can mimic prioritization within HA with automations in a creative way?

I created a new PR https://github.com/torbennehmer/hacs-e3dc/pull/155 which supports multiple wallboxes, so you may be able to use/test multiple ones soon.

Thomansky commented:
I will check it again tomorrow with the score. Unfortunately, my car is full, and the value has not been updated for 5 hours. I suspect that the 30 seconds come from the update rate that the integration sends to the house power plant. In the app Autarkie Manager, there is an option to set different update rates for power data and system parameters. Currently, I have an update rate of 10 seconds for power data. It might be a useful feature to add to the integration, allowing users to decide how often it updates. I apologize for my poor English; I am using a translator!
[image](https://github.com/torbennehmer/hacs-e3dc/assets/73141171/27bcb65b-e9ae-475e-b0a8-8b7438be8c64)

T1ppes commented:
I finally managed to update to the latest beta. Generally working and happy with the added features. I cannot provide any further feedback to the already mentioned. Although I currently have only one wallbox, I’d support to have the wallbox as a different device compared to the converter.
State of Charge provided by the wallbox does also not always align to the actual SoC of my car.
